### PR TITLE
satellite/overlay: add selected node cache

### PR DESCRIPTION
**Pull request by JessicaGreben**
Change-Id: Ia54a1943e0707a77189bc5f4a9aaa8339c98d99a

**What:** 

Add a cache that stores nodes that can be selected to upload data to.

This PR only implements the cache, there will be another PR to use with uploads.

**Why:**

When the uplink uploads a file, the FindStorageNode part of the upload is the slowest, so lets cache it to improve the speed.

**Please describe the tests:**
Add tests for cache.
 
**Please describe the performance impact:**
The selected-nodes-cache will get refreshed every 3 mins, this means that every replica of the satellite API will query the satellitedb's node table to get all of the nodes that qualify to store data on the network every 3 mins (this could occur less frequently).  This should be an acceptable amount of db load since there are currently only 4 API replicas running so it will only be 4 queries per 3 mins.  The `SelectAllStorageNodesUpload` db query executes in the slowest case (largest satellite) in about 20ms and returns 4951 rows.

## Code Review Checklist (to be filled out by reviewer)
 - [x] NEW: Are there any Satellite database migrations? Are they forwards _and_ backwards compatible? 
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

File: satellite/overlay/selectednodescache.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+				break
+			}
+		}
+	}
+
+	return selectedNodeResults, nil
+}
+
+// RemoveVetted removes a vetted node from the selected nodes cache
+func (c *SelectedNodesCache) RemoveVetted(ctx context.Context, nodeID storj.NodeID) {
```
[the remark this answers is not in the document]
Comment (JessicaGreben) [reply satellite/overlay/selectednodescache.go:140]: then in that case we would need to iterate thru both vetted and unvetted to find the node to rremove?

[the remark this answers is not in the document]
Comment (egonelbre) [reply satellite/overlay/selectednodescache.go:140]: If the caller doesn't know where it is, then it would need to call both RemoveVetted and RemoveUnvetted anyways.

File: satellite/overlay/selectednodescache.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+	LastIPPort     string
+	MinimumVersion string // semver or empty
+}
+
+// SelectedNodesCache keeps a list of all the storage nodes qualified to store data
+type SelectedNodesCache struct {
+	log              *zap.Logger
+	db               DB
+	vetted           []CachedNode
+	vettedMu         sync.Mutex
```
[the remark this answers is not in the document]
Comment (JessicaGreben) [reply satellite/overlay/selectednodescache.go:30]: what do you mean by "infrequent system updates"?

[the remark this answers is not in the document]
Comment (JessicaGreben) [reply satellite/overlay/selectednodescache.go:30]: also, im reading up on atomic value and sounds like it might be better to stick with mutexes... "These functions require great care to be used correctly" re: https://golang.org/pkg/sync/atomic/

Maybe we can start with mutexes and see if we need better performance. If so then consider using atomic value?

[the remark this answers is not in the document]
Comment (egonelbre) [reply satellite/overlay/selectednodescache.go]: Ssure, we can improve that later.

"These functions require great care to be used correctly" of course applies to mutexes, channels, goroutines as well.

File: satellite/overlay/config.go [hunk 3]
```diff
@@ -20,8 +20,9 @@ var (
 
 // Config is a configuration for overlay service.
 type Config struct {
-	Node                 NodeSelectionConfig
-	UpdateStatsBatchSize int `help:"number of update requests to process per transaction" default:"100"`
+	Node                        NodeSelectionConfig
+	UpdateStatsBatchSize        int           `help:"number of update requests to process per transaction" default:"100"`
+	SelectedNodesCacheStaleness time.Duration `help:"how stale the selected nodes cache can be" releaseDefault:"3m" devDefault:"5m"`
```
Comment (egonelbre) [inline satellite/overlay/config.go:25]: Make a separate config for cache.

File: satellite/satellitedb/overlaycache.go [hunk 4]
```diff
@@ -33,6 +33,80 @@ type overlaycache struct {
...
+// SelectAllStorageNodesUpload returns all nodes that qualify to store data, organized as reputable nodes and new nodes
+func (cache *overlaycache) SelectAllStorageNodesUpload(ctx context.Context, selectionCfg overlay.NodeSelectionConfig) (reputable, new []overlay.CachedNode, err error) {
+	defer mon.Task()(&ctx)(&err)
+	version, err := version.NewSemVer(selectionCfg.MinimumVersion)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	query := `
+		SELECT id, address, last_net, last_ip_port, true as reputable
```
Comment (egonelbre) [inline satellite/satellitedb/overlaycache.go:45]: You can do a single query without union all with `SELECT id, address, last_net, last_ip_port, (total_audit_count < $4 OR total_uptime_count < $5) as isnew`

File: satellite/overlay/service.go [hunk 5]
```diff
@@ -115,8 +118,11 @@ type NodeCheckInInfo struct {
 type FindStorageNodesRequest struct {
 	MinimumRequiredNodes int
 	RequestedCount       int
-	ExcludedIDs          []storj.NodeID
 	MinimumVersion       string // semver or empty
+	ExcludedIDs          []storj.NodeID
+	// TODO(jg): once the selected node cache is used in uploads
+	// remove the list of ExcludedIDs
+	ExcludedIDsMap map[storj.NodeID]struct{}
```
Comment (egonelbre) [inline satellite/overlay/service.go:125]: Keep using the slice for now.

File: satellite/overlay/selectednodescache.go [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+	db               DB
+	nodeSelectionCfg NodeSelectionConfig
+	staleness        time.Duration
+	lastRefresh      time.Time
+
+	reputableMu    sync.Mutex
+	reputableNodes []CachedNode
+
+	newNodesMu sync.Mutex
+	newNodes   []CachedNode
```
Comment (egonelbre) [inline satellite/overlay/selectednodescache.go:30]: Make a state struct similarly how reliability cache has. That way it's harder to make bugs related with update boundaries.

File: satellite/overlay/selectednodescache_test.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+		// confirm nodes are in the cache once initialized
+		err = cache.Init(ctx)
+		require.NoError(t, err)
+		reputable, new = cache.Size(ctx)
+		require.Equal(t, 2, reputable)
+		require.Equal(t, 0, new)
+	})
+}
+
+func TestRefresh(t *testing.T) {
```
Comment (egonelbre) [inline satellite/overlay/selectednodescache_test.go:59]: Add test for:
* concurrent GetNodes.
* concurrent GetNodes with low staleness.
* concurrent Refresh

File: satellite/overlay/selectednodescache.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+	"go.uber.org/zap"
+	"storj.io/common/storj"
+)
+
+// SelectedNodesCache keeps a list of all the storage nodes that are qualified to store data
+// We organize the nodes by if they are reputable or a new node on the network.
+// The cache will get refreshed once the staleness time has past.
+type SelectedNodesCache struct {
+	log              *zap.Logger
+	db               DB
```
Comment (egonelbre) [inline satellite/overlay/selectednodescache.go:21]: Use a separate database definition, it will helps to make some test writing easier.

File: satellite/overlay/selectednodescache.go [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+	"time"
+
+	"go.uber.org/zap"
+	"storj.io/common/storj"
+)
+
+// SelectedNodesCache keeps a list of all the storage nodes that are qualified to store data
+// We organize the nodes by if they are reputable or a new node on the network.
+// The cache will get refreshed once the staleness time has past.
+type SelectedNodesCache struct {
```
Comment (egonelbre) [inline satellite/overlay/selectednodescache.go:19]: `NodeSelectionCache`

File: satellite/overlay/selectednodescache.go [hunk 10]
```diff
@@ -0,0 +1,161 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package overlay
```
Comment (egonelbre) [inline satellite/overlay/selectednodescache.go:4]: Let's move this to a separate package `satellite/overlay/nodeselection/cache.go`, node selection will end up having quite a bit more logic in it.

File: satellite/overlay/selectednodescache.go [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+	log              *zap.Logger
+	db               DB
+	nodeSelectionCfg NodeSelectionConfig
+	staleness        time.Duration
+	lastRefresh      time.Time
+
+	reputableMu    sync.Mutex
+	reputableNodes []CachedNode
+
+	newNodesMu sync.Mutex
```
Comment (egonelbre) [inline satellite/overlay/selectednodescache.go:29]: You would need one mutex for the "state" and the other for the internal nodes themselves for removal.

File: satellite/overlay/selectednodescache.go [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+
+// SelectedNodesCache keeps a list of all the storage nodes that are qualified to store data
+// We organize the nodes by if they are reputable or a new node on the network.
+// The cache will get refreshed once the staleness time has past.
+type SelectedNodesCache struct {
+	log              *zap.Logger
+	db               DB
+	nodeSelectionCfg NodeSelectionConfig
+	staleness        time.Duration
+	lastRefresh      time.Time
```
Comment (egonelbre) [inline satellite/overlay/selectednodescache.go:24]: `lastRefresh` needs to be protected by something.

File: satellite/overlay/nodeselectioncache.go [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+	defer cache.mu.Unlock()
+	_, err = cache.refresh(ctx)
+	return err
+}
+
+// Refresh calls out to the database and refreshes the cache with the current data
+// from the nodes table, then sets time that the last refresh occurred so we know when
+// to refresh again in the future
+func (cache *NodeSelectionCache) refresh(ctx context.Context) (cachData *state, err error) {
+	defer mon.Task()(&ctx)(&err)
```
Comment (egonelbre) [inline satellite/overlay/nodeselectioncache.go:74]: This code is missing locking cache.mu.Lock. When two concurrent callers call "GetNodes" then this would race.

File: satellite/overlay/nodeselectioncache_test.go [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+	"context"
+	"strconv"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+	"storj.io/common/memory"
```
Comment (egonelbre) [inline satellite/overlay/nodeselectioncache_test.go:16]: import grouping needs  a fix

File: satellite/overlay/nodeselectioncache_test.go [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+				Timestamp:  time.Time{},
+				Release:    true,
+			},
+		}
+		err := db.UpdateCheckIn(ctx, n, time.Now().UTC(), nodeCfg)
+		require.NoError(t, err)
+	}
+}
+
+type mockdb struct {
```
Comment (egonelbre) [inline satellite/overlay/nodeselectioncache_test.go:91]: Add a capability to sleep inside `SelectAllStorageNodesUpload`.

File: satellite/overlay/nodeselectioncache_test.go [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+	mu        sync.Mutex
+	callCount int
+}
+
+func (m *mockdb) SelectAllStorageNodesUpload(ctx context.Context, selectionCfg overlay.NodeSelectionConfig) (reputable, new []*overlay.SelectedNode, err error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.callCount++
```
Comment (egonelbre) [inline satellite/overlay/nodeselectioncache_test.go:99]: time.Sleep(m.SleepDuration), here... it will make the

File: satellite/overlay/nodeselectioncache_test.go [hunk 17]
```diff
@@ -0,0 +1,255 @@
...
+	return []*overlay.SelectedNode{}, []*overlay.SelectedNode{}, nil
+}
+func TestRefreshConcurrent(t *testing.T) {
+	ctx := testcontext.New(t)
+
+	// concurrent refresh with high staleness
+	staleWhenAfter := time.Hour
+	mockDB := mockdb{}
```
Comment (egonelbre) [inline satellite/overlay/nodeselectioncache_test.go:107]: use a mock with sleep time.Second for concurrent tests. That way concurrent callers may see the stale state. This should end up detecting a race in the code as it is.

File: satellite/overlay/nodeselectioncache_test.go [hunk 18]
```diff
@@ -0,0 +1,255 @@
...
+		_, err := cache.GetNodes(ctx, overlay.FindStorageNodesRequest{})
+		return err
+	})
+	err := group.Wait()
+	require.NoError(t, err)
+	// expect only one call to GetNodes
+	require.Equal(t, 1, mockDB.callCount)
+
+	// concurrent get nodes with low staleness
+	staleWhenAfter = time.Nanosecond
```
Comment (egonelbre) [inline satellite/overlay/nodeselectioncache_test.go:214]: Use a negative or zero value here.

All platforms can have the behavior of, however it's the worst on Windows:
```
a := time.Now()
b := time.Now()
fmt.Println(a.Equal(b)) // prints true
```

File: satellite/overlay/nodeselectioncache_test.go [hunk 19]
```diff
@@ -0,0 +1,255 @@
...
+}
+
+func (m *mockdb) SelectAllStorageNodesUpload(ctx context.Context, selectionCfg overlay.NodeSelectionConfig) (reputable, new []*overlay.SelectedNode, err error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.callCount++
+	return []*overlay.SelectedNode{}, []*overlay.SelectedNode{}, nil
+}
+func TestRefreshConcurrent(t *testing.T) {
+	ctx := testcontext.New(t)
```
Comment (egonelbre) [inline satellite/overlay/nodeselectioncache_test.go:103]: `testcontext.New` requires `ctx.Cleanup` to be called.

File: satellite/satellitedb/overlaycache.go [hunk 20]
```diff
@@ -33,6 +33,71 @@ type overlaycache struct {
 	db *satelliteDB
 }
 
+// SelectAllStorageNodesUpload returns all nodes that qualify to store data, organized as reputable nodes and new nodes
+func (cache *overlaycache) SelectAllStorageNodesUpload(ctx context.Context, selectionCfg overlay.NodeSelectionConfig) (reputable, new []*overlay.SelectedNode, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	if selectionCfg.MinimumVersion == "" {
```
Comment (egonelbre) [inline satellite/satellitedb/overlaycache.go:40]: As I mentioned, changing which nodes are selected should be a separate PR either before or after this PR.

However, changing this would require changing quite a lot of code due to storj-sim, backwards compatibility tests etc.

File: satellite/overlay/nodeselectioncache_test.go [hunk 21]
```diff
@@ -0,0 +1,255 @@
...
+			require.NotEqual(t, node.LastNet, "")
+		}
+	})
+}
+func TestGetNodeConcurrent(t *testing.T) {
+	ctx := testcontext.New(t)
+
+	// concurrent GetNodes with high staleness
+	staleWhenAfter := time.Hour
+	mockDB := mockdb{}
```
Comment (egonelbre) [inline satellite/overlay/nodeselectioncache_test.go:192]: Use a mock with sleep here to ensure both GetNodes see the staleness.

File: satellite/overlay/nodeselectioncache_test.go (hunk 16 above)
Comment (JessicaGreben) [reply satellite/overlay/nodeselectioncache_test.go:99]: "it will make the" what?  why do we want the ability to sleep in a tests?

Comment (egonelbre) [reply satellite/overlay/nodeselectioncache_test.go:99]: Ah, sorry, I forgot to finish the sentence. It will make sure that the database call doesn't happen too quickly and other callers can see the stale state.

File: satellite/overlay/nodeselectioncache.go [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+		db:              db,
+		staleness:       staleness,
+		selectionConfig: config,
+		data:            &state{},
+	}
+}
+
+// Init populates the cache with all of the reputableNodes and newNode nodes
+// that qualify to upload data from the nodes table in the overlay database
+func (cache *NodeSelectionCache) Init(ctx context.Context) (err error) {
```
Comment (egonelbre) [inline satellite/overlay/nodeselectioncache.go:62]: Oh, I just realized that you don't even need the init, since during the first call to GetNodes it will update the cache.

File: satellite/satellitedb/overlaycache.go [hunk 23]
```diff
@@ -33,6 +33,72 @@ type overlaycache struct {
...
+			FROM nodes
+			WHERE disqualified IS NULL
+			AND suspended IS NULL
+			AND exit_initiated_at IS NULL
+			AND type = $3
+			AND free_disk >= $4
+			AND last_contact_success > $5
+	`
+	args := []interface{}{
+		// $1, $2
```
Comment (navillasa) [inline satellite/satellitedb/overlaycache.go:51]: nice

File: satellite/satellitedb/overlaycache.go [hunk 24]
```diff
@@ -33,6 +33,72 @@ type overlaycache struct {
...
+			return nil, nil, err
+		}
+		query += `AND (major > $6 OR (major = $7 AND (minor > $8 OR (minor = $9 AND patch >= $10)))) AND release`
+		args = append(args,
+			// $6 - $10
+			version.Major, version.Major, version.Minor, version.Minor, version.Patch,
+		)
+	}
+
+	rows, err := cache.db.Query(ctx, query, args...)
```
Comment (navillasa) [inline satellite/satellitedb/overlaycache.go:72]: I think this needs `defer func() { err = errs.Combine(err, rows.Close()) }()` after the err check block

File: satellite/satellitedb/overlaycache.go [hunk 25]
```diff
@@ -33,6 +33,72 @@ type overlaycache struct {
...
+		}
+
+		if isnew {
+			newNodes = append(newNodes, &node)
+			continue
+		}
+		reputableNodes = append(reputableNodes, &node)
+	}
+
+	return reputableNodes, newNodes, nil
```
Comment (navillasa) [inline satellite/satellitedb/overlaycache.go:99]: I think the last return arg should be `Error.Wrap(rows.Err())`

File: satellite/overlay/nodeselectioncache_test.go [hunk 26]
```diff
@@ -0,0 +1,273 @@
...
+		// the cache should have no nodes to start
+		reputable, new := cache.Size()
+		require.Equal(t, 0, reputable)
+		require.Equal(t, 0, new)
+
+		// add some nodes to the database
+		const nodeCount = 4
+		addNodesToNodesTable(ctx, t, db.OverlayCache(), nodeCount)
+
+		// confirm cache.GetNodes returns the correct nodes
```
Comment (navillasa) [inline satellite/overlay/nodeselectioncache_test.go:187]: suggestion for future PR: there's a lot of existing tests around correct selection in satellite/overlay/selection_test.go and TestRandomizedSelection in satellite/overlay/service_test.go -- I think those will eventually need to be repurposed to test selection from this cache.

File: satellite/overlay/nodeselectioncache.go [hunk 27]
```diff
@@ -0,0 +1,195 @@
...
+	// how many reputableNodes versus newNode nodes should be selected
+	totalcount := req.RequestedCount
+	newNodeCount := int(float64(req.RequestedCount) * newNodeFraction)
+
+	var selectedNodeResults = []*SelectedNode{}
+	var distinctNetworks = map[string]struct{}{}
+
+	// Get a random selection of new nodes out of the cache first so that if there aren't
+	// enough new nodes on the network, we can fall back to using reputable nodes instead
+	randomIdexes := rand.Perm(len(cacheData.newNodes))
```
[the remark this answers is not in the document]
Comment (JessicaGreben) [reply satellite/overlay/nodeselectioncache.go:134]: oops good catch. _fixing_

File: satellite/overlay/nodeselectioncache_test.go (hunk 26 above)
Comment (JessicaGreben) [reply satellite/overlay/nodeselectioncache_test.go:187]: when we plug this cache into the code to use with uploads, we will update tests

File: satellite/overlay/nodeselectioncache.go (hunk 22 above)
Comment (JessicaGreben) [reply satellite/overlay/nodeselectioncache.go:62]: i think it might be nice to init the cache when the api starts up. but i could be talked in to only doing it during GetNodes. although i would like to try to reduce the number of times we refresh from within the GetNodes method since its blocking. thoughts?

File: satellite/overlay/nodeselectioncache.go [hunk 28]
```diff
@@ -0,0 +1,195 @@
...
+// If the cache hasn't been refreshed recently it will do so first.
+func (cache *NodeSelectionCache) GetNodes(ctx context.Context, req FindStorageNodesRequest) (_ []*SelectedNode, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	cache.mu.RLock()
+	cacheData := cache.data
+	cache.mu.RUnlock()
+
+	// if the cache is stale, then refresh it before we get nodes
+	if time.Since(cacheData.lastRefresh) > cache.staleness {
```
[the remark this answers is not in the document]
Comment (JessicaGreben) [reply satellite/overlay/nodeselectioncache.go:106]: it is a little redundant and we prob don't need this block, we could just check in the refresh method, but i do like the way it reads better. Without this if block, at first glance, it would look like the GetNodes method always refreshes the cache which would be bad

File: satellite/overlay/nodeselectioncache.go (hunk 22 above)
Comment (egonelbre) [reply satellite/overlay/nodeselectioncache.go:62]: I mean, it might take some time from init and to first call to GetNodes... so it might do that anyways. I don't think it will make a perf difference, however, I would remove this mainly because it will be less code.

File: satellite/overlay/nodeselectioncache.go (hunk 28 above)
[the remark this answers is not in the document]
Comment (egonelbre) [reply satellite/overlay/nodeselectioncache.go:106]: No, we definitely need this check. It's this way to avoid using write lock in the common case.

File: satellite/overlay/nodeselectioncache.go (hunk 22 above)
Comment (JessicaGreben) [reply satellite/overlay/nodeselectioncache.go]: ok i removed Init method, but replaced with public Refresh method to use in tests

File: satellite/overlay/nodeselectioncache.go (hunk 28 above)
[the remark this answers is not in the document]
Comment (navillasa) [reply satellite/overlay/nodeselectioncache.go:106]: oh okay, good to know
